### PR TITLE
added second framework to watch pods

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards https://github.com/giantswarm/giantswarm/issues/1918 and [the design slides document](https://docs.google.com/presentation/d/1810pJqZVkFvgCpJ4WpBkSfCpIizUk0550Re8xA1RDPQ/edit#slide=id.g29694f039e_0_363). See also https://github.com/giantswarm/giantswarm/pull/1889. Note this is based on the `draining-nodes` branch.

### CODE REVIEW

File: service/framework.go [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+	podresource "github.com/giantswarm/kvm-operator/service/resource/pod"
+	pvcresource "github.com/giantswarm/kvm-operator/service/resource/pvc"
+	serviceresource "github.com/giantswarm/kvm-operator/service/resource/service"
+)
+
+const (
+	ResourceRetries uint64 = 3
+)
+
+func newCustomObjectFramework(config Config) (*framework.Framework, error) {
```
Comment (xh3b4sd) [inline service/framework.go]: There are now multiple factories for operatorkit frameworks.

File: service/framework.go [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+		}
+	}
+
+	initCtxFunc := func(ctx context.Context, obj interface{}) (context.Context, error) {
+		ctx = messagecontext.NewContext(ctx, messagecontext.NewMessage())
+
+		return ctx, nil
+	}
+
+	var customObjectFramework *framework.Framework
```
Comment (xh3b4sd) [inline service/framework.go]: All this is basically copy/paste from `service/service.go`.

File: service/framework.go [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+		customObjectFramework, err = framework.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	return customObjectFramework, nil
+}
+
+func newPodFramework(config Config) (*framework.Framework, error) {
```
Comment (xh3b4sd) [inline service/framework.go]: This is the new additional pod framework. We make the kvm-operator able to watch multiple different resources. We want to make the operator also watching pods it creates itself. That way we can manage the deletion of the guest cluster pods within separated resource implementations and do the draining magic.

File: service/framework.go [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+		c.TLS.CrtFile = config.Viper.GetString(config.Flag.Service.Kubernetes.TLS.CrtFile)
+		c.TLS.KeyFile = config.Viper.GetString(config.Flag.Service.Kubernetes.TLS.KeyFile)
+
+		k8sClient, err = k8sclient.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var podResource framework.Resource
```
Comment (xh3b4sd) [inline service/framework.go]: The pod resource is new and will manage the guest cluster node draining.

File: service/resource/pod/current.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	currentPod, err := toPod(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	fmt.Printf("\n")
+	fmt.Printf("pod resource start: TODO reconciling pod\n")
+	fmt.Printf("%#v\n", currentPod)
+	fmt.Printf("pod resource end: TODO reconciling pod\n")
+	fmt.Printf("\n")
```
Comment (xh3b4sd) [inline service/resource/pod/current.go]: I am going to implement this here. I saw in the logs that pods having the right label will be reconciled here. One of the next steps is to add the right label to the guest cluster worker pods.

File: service/resource/pod/resource.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+		Logger:    nil,
+	}
+}
+
+type Resource struct {
+	k8sClient kubernetes.Interface
+	logger    micrologger.Logger
+}
+
+func New(config Config) (*Resource, error) {
```
Comment (xh3b4sd) [inline service/resource/pod/resource.go]: This is the new pod resource. Basically only structure. Implementation will follow separately.

File: service/service.go [hunk 7]
```diff
@@ -112,227 +75,30 @@ func New(config Config) (*Service, error) {
...
-	var certWatcher certificatetpr.Searcher
-	{
-		certConfig := certificatetpr.DefaultServiceConfig()
-		certConfig.K8sClient = k8sClient
-		certConfig.Logger = config.Logger
-		certWatcher, err = certificatetpr.NewService(certConfig)
-		if err != nil {
-			return nil, microerror.Mask(err)
-		}
-	}
```
Comment (xh3b4sd) [inline service/service.go]: All this moved.

File: service/framework.go [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+	}
+
+	var newInformer *informer.Informer
+	{
+		c := informer.DefaultConfig()
+
+		c.WatcherFactory = newWatcherFactory
+
+		c.RateWait = 10 * time.Second
+		c.ResyncPeriod = 5 * time.Minute
```
Comment (kopiczko) [inline service/framework.go]: Why not using the defaults?

Comment (xh3b4sd) [reply service/framework.go]: No special reason. I think the values will change. I just do not know yet which values are better and why. Lets learn and adapt along the road.

Comment (kopiczko) [reply service/framework.go]: FYI: My thinking is it's nice to have all defaults inside operatorkit, and change it only when we have a particular reason for that. That gives us ability to change the default globally.

**Pull request merged**